$OpenBSD: patch-salt_tokens_localfs_py,v 1.1 2020/05/01 08:01:08 robert Exp $

commit 7bd0ab195fbec4f34523dad11149f741c154e2b7
Author: Daniel A. Wozniak <dwozniak@saltstack.com>
Date:   Mon Apr 13 06:44:58 2020 +0000

    Fix CVE-2020-11652
    
    Sanitize paths in ClearFuncs methods provided by salt-master. This
    ensures we do not allow access to un-intended files and directories.

Index: salt/tokens/localfs.py
--- salt/tokens/localfs.py.orig
+++ salt/tokens/localfs.py
@@ -12,6 +12,7 @@ import logging
 
 import salt.utils.files
 import salt.utils.path
+import salt.utils.verify
 import salt.payload
 
 from salt.ext import six
@@ -34,6 +35,8 @@ def mk_token(opts, tdata):
     hash_type = getattr(hashlib, opts.get('hash_type', 'md5'))
     tok = six.text_type(hash_type(os.urandom(512)).hexdigest())
     t_path = os.path.join(opts['token_dir'], tok)
+    if not salt.utils.verify.clean_path(opts['token_dir'], t_path):
+        return {}
     while os.path.isfile(t_path):
         tok = six.text_type(hash_type(os.urandom(512)).hexdigest())
         t_path = os.path.join(opts['token_dir'], tok)
